heron/ckptmgr/src/java/com/twitter/heron/ckptmgr/CheckpointManagerServer.java

```diff
@@ -0,0 +1,297 @@
+import com.twitter.heron.common.network.HeronSocketOptions;
+import com.twitter.heron.common.network.REQID;
+import com.twitter.heron.proto.ckptmgr.CheckpointManager;
+import com.twitter.heron.proto.system.Common;
+
+public class CheckpointManagerServer extends HeronServer {
+  private static final Logger LOG = Logger.getLogger(CheckpointManagerServer.class.getName());
+
+  private final String topologyName;
+  private final String topologyId;
+  private final String checkpointMgrId;
+  private final IBackend checkpointsBackend;
+
+  private SocketChannel connection;
+
+  public CheckpointManagerServer(
+      String topologyName, String topologyId, String checkpointMgrId,
+      IBackend checkpointsBackend, NIOLooper s, String host,
+      int port, HeronSocketOptions options) {
+    super(s, host, port, options);
+
+    this.topologyName = topologyName;
+    this.topologyId = topologyId;
+    this.checkpointMgrId = checkpointMgrId;
+    this.checkpointsBackend = checkpointsBackend;
+
+    this.connection = null;
+
+    registerInitialization();
+  }
+
+  private void registerInitialization() {
+    registerOnRequest(CheckpointManager.RegisterStMgrRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.RegisterTMasterRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.SaveInstanceStateRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.GetInstanceStateRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.CleanStatefulCheckpointRequest.newBuilder());
+  }
+
+  @Override
+  public void onConnect(SocketChannel channel) {
+    LOG.info("Got a new connection from host:port "
+        + channel.socket().getRemoteSocketAddress());
+  }
+
+  @Override
+  public void onRequest(REQID rid, SocketChannel channel, Message request) {
+    if (request instanceof CheckpointManager.RegisterStMgrRequest) {
+      handleStMgrRegisterRequest(rid, channel, (CheckpointManager.RegisterStMgrRequest) request);
+    } else if (request instanceof CheckpointManager.RegisterTMasterRequest) {
+      handleTMasterRegisterRequest(rid, channel,
+                                   (CheckpointManager.RegisterTMasterRequest) request);
+    } else if (request instanceof CheckpointManager.SaveInstanceStateRequest) {
+      handleSaveInstanceStateRequest(
+          rid, channel, (CheckpointManager.SaveInstanceStateRequest) request);
+    } else if (request instanceof CheckpointManager.GetInstanceStateRequest) {
+      handleGetInstanceStateRequest(
+          rid, channel, (CheckpointManager.GetInstanceStateRequest) request);
+    } else if (request instanceof CheckpointManager.CleanStatefulCheckpointRequest) {
+      handleCleanStatefulCheckpointRequest(
+          rid, channel, (CheckpointManager.CleanStatefulCheckpointRequest) request);
+    } else {
+      LOG.severe("Unknown kind of request: " + request.getClass().getName());
+    }
+  }
+
+  protected void handleCleanStatefulCheckpointRequest(
```

Comment:
a.) The current approach allows us to share code via composition. Using a handler interface allows that as well as via polymorphism or the decorator pattern, the filter pattern or the template pattern, which can be useful in request handling.
c.) Note the "more modular" part.
d.) Overtime this class will continue to grow in size as each RPC develops more functionality and bug fixes and as we add more RPC methods. It's generally good practice to favor smaller classes that support a specific purpose when possible, versus combining multiple concerns into one. Even with it's current size it would be easier to understand the codebase if you could browse handler classes named after their RPC call. Hence putting good patterns in place now will help prevent bloat. Also see https://en.wikipedia.org/wiki/God_object
e) Using the handler pattern won't add coupling to other backend functionality. If you're concerned about mixing RPC classes with other unrelated classes, then we could put the RPC/server related code in it's own subpackage.